### DISCUSSION TITLE
Add a new HTTP engine that supports HTTP/2

### DISCUSSION THREAD

**Discussion by ikhoon** [playframework/playframework#12004] [Ideas]
HTTP/2 is an efficient protocol for performance and it is wildly used in many places.
However, Netty server in Playframework does not support HTTP/2. 
Akka HTTP is the only choice for HTTP/2 but it is still in the [preview](https://doc.akka.io/docs/akka-http/current/server-side/http2.html) stage.
It would be useful if users can choose a stable HTTP/2 engine out of the box.

[Armeria](https://armeria.dev/) is a light-weight HTTP/1 and HTTP/2 server built top of Netty.
If the maintainers are welcome to adopt Armeria as Playframework's another HTTP engine,
I'd like to send a PR for it. 😀

**Comment by mkurz** [comment]
Actually it should be doable to add another server backend. You find the current backends here: https://github.com/playframework/playframework/tree/main/transport/server That should give you a good starting point on how to add the new backend. 
Be aware to implement all the configs that are present in the play-server project: https://github.com/playframework/playframework/blob/main/transport/server/play-server/src/main/resources/reference.conf These are general settings, supported by both the netty and the akka-http backend, and any new backend should implement them as well. Any specific settings can then go in a specific reference.conf, like for netty see: https://github.com/playframework/playframework/blob/main/transport/server/play-netty-server/src/main/resources/reference.conf

**Comment by ikhoon** [comment]
Thanks for the guidance. I will start implementing the Armeria server backend and send a PR when it is ready. 😀

**Comment by tmccombs** [comment]
Netty also has support for http2. Play just doesn't currently use it (yet).

**Reply by tmccombs** [reply]
I briefly started working on a branch to add http2 support with netty at https://github.com/tmccombs/playframework/tree/http2-netty

It is very much a work in progress, but I haven't had a lot of time to spend on it. If someone wanted to pick it up, feel free to build on what I have so far.
